File: LeetCode 30 Day Challenge/15-productOfArrayExceptSelf.py
# N is length of nums array
# Time Complexity: O(N)
# Space Complexity: O(1) excluding the output array

class Solution:
    def productExceptSelf(self, nums):
        output = [1] * len(nums)
        left = 1
        right = 1
        for i in range(len(nums)):
            output[i] *= left
            output[-1-i] *= right
            left *= nums[i]
            right *= nums[-1-i]
        return output

File: LeetCode 30 Day Challenge/15-productOfArrayExceptSelf.js
// N is length of nums array
// Time Complexity: O(N)
// Space Complexity: O(1) excluding the output array

const productExceptSelf = (nums) => {
    const output = new Array(nums.length).fill(1)
    let left = 1
    for (let i = 0; i < nums.length; i++) {
        output[i] *= left
        left *= nums[i]
    }
    let right = 1
    for (let i = nums.length - 1; i >= 0; i--) {
        output[i] *= right
        right *= nums[i]
    }

    return output
}